Update with standard and sidecar modes (#26)
datastax/vault-plugin-secrets-datastax-astra: docs/index.md

Astra DB Plugin for HashiCorp Vault is offered under the open-source [Apache 2.0](../LICENSE.txt) license.

## Available caller modes

> Astra DB Plugin for HashiCorp Vault supports two caller modes, which determine the behavior of the plugin when creating tokens:

Reviewer: Is there a reason you didn't use the attribute {astra_db} instead of Astra DB?

* *Standard*: Designed for applications that call the API directly without using a Vault agent. This mode is the default behavior.

* *Sidecar*: Designed for Vault agents running in a sidecar container. In this mode, the plugin API behaves as described in the [Vault Sidecar agent](https://developer.hashicorp.com/vault/docs/platform/k8s/injector) documentation. 

The major differences between the standard and sidecar caller modes:

> * In `standard` mode, the `vault write` command is used to generate new tokens. The `sidecar` mode does not support the `vault write` command.

Reviewer: ```suggestion
* In `standard` mode, the `vault write` command generates new tokens. The `sidecar` mode does not support the `vault write` command.
```

> * In `standard` mode, the `vault read` command is used to read a token; that is, get an existing token's information from vault. The `sidecar` mode uses the `vault read` command to generate new tokens. 

Reviewer: ```suggestion
* In `standard` mode, the `vault read` command gets an existing token's information from vault (also known as reading a token). The `sidecar` mode uses the `vault read` command to generate new tokens. 
```

Astra DB Plugin for HashiCorp Vault is free. See the HashiCorp Platform Vault site for its [enterprise pricing](https://cloud.hashicorp.com/products/vault/pricing) details. 

<a name="build-steps"></a>
## Build steps - optional

> If you elect to build the plugin from Go modules in our GitHub repo, follow these steps. Otherwise, you can use the provided binary.

Reviewer: ```suggestion
To build the plugin from Go modules in our GitHub repo, follow these steps. Otherwise, you can use the provided binary.
```

Author: We want to emphasize that building from source is optional.  I'll edit this part separately in a later commit.

Author: Edited in subsequent commit.

2. Enable the plugin in your Vault instance:

	```bash
	vault secrets enable -path=astra vault-plugin-secrets-datastax-astra
	```

	*Output*:

	```bash
	Success! Enabled the vault-plugin-secrets-datastax-astra secrets engine at: astra/
	```

3. In this step, configure the credentials Vault uses to generate tokens. Add a root token for each Astra DB organization. Specify the caller mode of the plugin (`standard` or `sidecar`). Standard is the default.

    Standard caller mode example:

    ```bash
    vault write astra/config org_id="<ORG ID>" astra_token="<YOUR ASTRA ADMINISTRATOR APP TOKEN>" 
      url="https://api.astra.datastax.com" logical_name="<YOUR LOGICAL NAME>" caller_mode="standard"
    ```

    Sidecar caller mode example:

    ```bash
    vault write astra/config org_id="<ORG ID>" astra_token="<YOUR ASTRA ADMINISTRATOR APP TOKEN>" 
      url="https://api.astra.datastax.com" logical_name="<YOUR LOGICAL NAME>" caller_mode="sidecar"
    ```

    *Output*:

    ```bash
    Success! Data written to astra/configs
    ```

## Setup plugin from binary distribution

> 1. Create a plugins directory where HashiCorp Vault will find the plugin. Example: `./vault/plugins`.  **IMPORTANT:** do not specify a symlinked directory.

Reviewer: ```suggestion
1. Create a plugin's directory where HashiCorp Vault finds the plugin. Example: `./vault/plugins`.  
+
**IMPORTANT** 
====
Do not specify a symlinked directory.
====
```

> At this point, HashiCorp Vault and Astra DB Plugin for HashiCorp Vault are set up. They're ready to use.

Reviewer: ```suggestion
At this point, HashiCorp Vault and Astra DB Plugin for HashiCorp Vault are set up and ready to use.
```

* A wholesale org
* An internal usage org

> *IMPORTANT*: As noted previously in this topic, here are the differences between `standard` and `sidecar` caller modes:

Reviewer: Is there a reason this isn't written as a normal admonition?

Author: It's markdown...

> * In `standard` mode, the `vault write` command is used to generate new tokens. The `sidecar` mode does not support the `vault write` command.

Reviewer: ```suggestion
* In `standard` mode, the `vault write` command generates new tokens. The `sidecar` mode does not support the `vault write` command.
```

> * In `standard` mode, the `vault read` command is used to read a token; that is, get an existing token's information from vault. The `sidecar` mode uses the `vault read` command to generate new tokens. 

Reviewer: ```suggestion
* In `standard` mode, the `vault read` command gets an existing token's information from vault (also known as reading a token). The `sidecar` mode uses the `vault read` command to generate new tokens. 
```

> 	The created root token will be used by HashiCorp Vault for further token operations within this organization. 

Reviewer: ```suggestion
	HashiCorp Vault then uses the created root token for further token operations within this organization. 
```

> 	config/Y3tAnotherOrgId_aaa777777_bbbb_4ad3_ccc77777d

Reviewer 2: In the latest version, the list will only display the orgIds, there won't be 'config/`f

	 --header 'Authorization: Bearer <application_token>'
	```

> 	Or you can run the [update_roles.sh](https://github.com/datastax/vault-plugin-secrets-datastax-astra/blob/main/update_roles.sh) script. It's provided in our GitHub repo. The script adds all the Astra DB roles (default and custom) and their IDs to HashiCorp Vault. The update_roles.sh script generates Vault roles from Astra DB roles with default `ttl` and `max_ttl` values (86400 seconds). Example:

Reviewer: ```suggestion
	You can also run the [update_roles.sh](https://github.com/datastax/vault-plugin-secrets-datastax-astra/blob/main/update_roles.sh) script, available in our GitHub repo. The script adds all the Astra DB roles (default and custom) and their IDs to HashiCorp Vault. The update_roles.sh script generates Vault roles from Astra DB roles with default `ttl` and `max_ttl` values (86400 seconds). Example:
```

	```

> 	Also available is the `vault delete astra/role org_id="<ORG ID>" role="<ROLE NAME>"` command.

Reviewer: ```suggestion
	The `vault delete astra/role org_id="<ORG ID>" role="<ROLE NAME>"` command is also available.
```

    metadata           <nil>
    orgId              ccd999999_facd_4ad3_bbb99903d999999999999999d
    roleName           r_w_user
    token              AstraCS:DTEUyKUZZrvFzzLzPGCbrbHO:4ebdef6dbbcada2f4618c9e4b6800c17b10c...
    ```

    After the lease expires, the token is automatically revoked; it's deleted from the Vault and from Astra DB.

9. To read a generated token using `client_id` and `org_id`:

    ```bash
    vault read astra/org/token client_id="ZenclHmfKBenxQZjRNhRUbLc" org_id="ccd999999_facd_4ad3_bbb99903d999999999999999d"
    ```
    
    *Output*:

    ```bash
    clientId       ZenclHmfKBenxQZjRNhRUbLc
    generatedOn    2022-09-21T17:40:29.265975403Z
    logicalName    org1
    metadata       <nil>
    orgId          ccd999999_facd_4ad3_bbb99903d999999999999999d
    roleName       r_w_user
    token          AstraCS:ZenclHmfKBenxQZjRNhRUbLc:a4bf8ca1bf164698e2631d353a295be27187e5750afc4080a4749fcc68f99f38
    ```

10. To read a token using `org_id`, `role_name`, and `logical_name`:

    ```bash
    vault read astra/org/token org_id=”ccd999999_facd_4ad3_bbb99903d999999999999999d” \
    role_name=”r_w_user” logical_name=”org1”
    ```

    *Output*:

    ```bash

>     clientId       ZenclHmfKBenxQZjRNhRUbLc

Reviewer: Did you change any of the token or orgIDs just in case someone tries to use them?

Author: Yes, they're outdated or fake.

    ```

11. To list all Astra DB tokens:

    ```bash
    vault list astra/org/tokens
    ```

    *Output*:

    ```bash
    CtQUkZlxYDzoxydARnbhjlji
    YZZcEggpHvLPZQnnioYbZqLc
    rGQajQDqoPNyjtiokNNrIgPK
    ```

12. To force revoke a token before the lease expires:

    ```bash
    vault lease revoke <lease_id>
    ```

    The token is revoked; it's deleted from the Vault and from Astra DB.

13. To renew a token before the lease expires:

	```bash
    vault lease renew -increment=3600 <lease_id>
	```

	*Output*:

	```bash
    Key                Value
    ---                -----
    lease_id           astra/org/token/ove4ll9sshXFL3qK6I1qPezb
    lease_duration     1h
    lease_renewable    true
	```

    *NOTE*: The token can be renewed to a value no greater than `max_ttl`.

<a name="using-sidecar"></a>
## Sidecar caller mode - using Astra DB tokens with HashiCorp Vault

This section assumes you're using the plugin binary or you have optionally built the plugin from source (optional), using the steps described earlier in [this topic](#build-steps).

*REMINDER*: As previously noted:

> * In `standard` mode, the `vault write` command is used to generate new tokens. The `sidecar` mode does not support the `vault write` command.

Reviewer: ```suggestion
* In `standard` mode, the `vault write` command generates new tokens. The `sidecar` mode does not support the `vault write` command.
```

> * In `standard` mode, the `vault read` command is used to read a token; that is, get an existing token's information from vault. The `sidecar` mode uses the `vault read` command to generate new tokens. 

Reviewer: ```suggestion
* In `standard` mode, the `vault read` command gets an existing token's information from vault (also known as reading a token). The `sidecar` mode uses the `vault read` command to generate new tokens. 
```

    vault lease renew -increment=3600 <lease_id>
	```

	*Output*:

	```bash
    Key                Value
    ---                -----
    lease_id           astra/org/token/ove4ll9sshXFL3qK6I1qPezb
    lease_duration     1h
    lease_renewable    true
	```

    *NOTE*: The token can be renewed to a value no greater than `max_ttl`.

## Upgrade plugin from binary distribution

Plugin versioning was introduced in Vault 1.12, allowing for a smooth upgrade of the plugin that has been mounted at a path on a running Vault server. These steps assume you have already registered the plugins as outlined under "Setup plugin from binary distribution".

1. To upgrade the plugin, register a newer version of the plugin. You **must** use the same plugin type (secret) and name (vault-plugin-secrets-datastax-astra) as the plugin being upgraded:

	```bash
	vault plugin register -sha256=${SHA256} \
	  -command=vault-plugin-secrets-datastax-astra_1.0.1 \
	  -version=v1.0.1 \
	  secret vault-plugin-secrets-datastax-astra
	```

2. Tune the existing mount to configure it to use the newly registered version:

	```bash
	vault secrets tune -plugin-version=v1.0.1 vault-plugin-secrets-datastax-astra
	```

> 3. If you want, you can check the updated configuration. Notice the "Version" is now different from the "Running Version":

Reviewer: ```suggestion
3. You can check the updated configuration. Notice the "Version" is now different from the "Running Version":
```

	```bash
	vault secrets list -detailed
	```

4. As the final step, trigger a plugin reload. This will reload all mounted backends using that plugin:

	```bash
	vault plugin reload -plugin vault-plugin-secrets-datastax-astra
	```

## Summary

> HashiCorp Vault has a full understanding of the historical token specifics, for control and auditing purposes, including when the tokens were used and by whom, along with a free-form role name and any custom metadata you may have associated with the tokens. For example, HashiCorp Vault's data knows the details of the token delete operation via its identity management and access control data; whereas Astra DB is only aware that a token of a particular `clientId` was generated on a date, and has since been deleted.

Reviewer: ```suggestion
HashiCorp Vault has a full understanding of the historical token specifics, for control and auditing purposes, including when the tokens were used and by whom, along with a free-form role name and any custom metadata you may have associated with the tokens. For example, HashiCorp Vault's data knows the details of the token delete operation through its identity management and access control data; whereas Astra DB is only aware that a token of a particular `clientId` was generated on a date, and has since been deleted.
```

> 	vault read astra/role org_id="<ORG ID>" role="<ROLE NAME>"

Reviewer 2: role_name instead of role
